### PR TITLE
2.3-free-monoids.md #14

### PR DESCRIPTION
**Pull request by arioston**
Pre Work 2.3-free-monoids.md #14

### CODE REVIEW

File: src/main/ank/2.3-free-monoids.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  override fun mempty() = listOf<M>()
+  override fun mappend(a1: List<M>, a2: List<M>) = a1 + a2
+}
+```
+................
+```Haskell
+2 * 3 = 6
+[2] ++ [3] = [2, 3] // not the same as [6]
+```
+```kotlin:ank:silent
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]: Change this to kotlin, as this is not intended to be compiled. 
I will add more explanations later to the README.

File: src/main/ank/2.3-free-monoids.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+listOf(2) + listOf(3) == listOf(2, 3)
+```
+
+................
+```Haskell
+2 * 3 = 6
+```
+```kotlin:ank:silent
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]: change to kotlin

File: src/main/ank/2.3-free-monoids.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+```
+```kotlin
+val h: (M) -> N
+```
+
+................
+```Haskell
+q = U h . p
+```
+```kotlin
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md:89]: U is a Functor in your snippet it could be a variable uh. You may change it to U(h).

**Review by i-walker — COMMENTED** [review]
Don't forget your imports.

**Comment by i-walker** [comment]
Feel free to add more commits and I will help you along the way until you're done with the chapter. 
With that, I mean to integrate it with Arrow's [Free](https://github.com/arrow-kt/arrow/blob/master/modules/free/arrow-free-data/src/main/kotlin/arrow/free/Free.kt).

We can also work on it together (I am new to Free, too) here are some of the resources I found so far.
[A nice'ish explaination](https://stackoverflow.com/questions/13352205/what-are-free-monads)
[In the Kotlin slack](https://kotlinlang.slack.com/archives/C5UPMM0A0/p1554022508033000)

File: src/main/ank/2.3-free-monoids.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+}
+```
+................
+```Haskell
+instance Monoid [a] where
+  mempty = []
+  mappend = (++)
+```
+```kotlin:ank:silent
+class ListMonoid<M> : Monoid<List<M>> {
```
Comment (raulraja) [inline src/main/ank/2.3-free-monoids.md]: Use `A` as a type arg instead of `M` since that is the Arrow convention for concrete not kinded type args

**Comment by arioston** [comment]
@i-walker Thanks, I was trying to figure out how to incorporate Arrow's Free.  I will appreciate if we can work together.

**Comment by i-walker** [comment]
@arioston If you have any questions whatsoever don't hesitate to ask in the #arrow-contributor slack group. 
> @i-walker Thanks, I was trying to figure out how to incorporate Arrow's Free. I will appreciate if we can work together.

File: src/main/ank/2.3-free-monoids.md [hunk 5]
```diff
@@ -4,9 +4,20 @@ class Monoid m where
   mappend :: m -> m -> m
 ```
 ```kotlin:ank:silent
-interface Monoid<M> {
-  fun mempty() : M
-  fun mappend(a1: M, a2: M): M
+package arrow.typeclasses
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]: delete this line, it's unnecessary.
>  package arrow.typeclasses

File: src/main/ank/2.3-free-monoids.md [hunk 6]
```diff
@@ -4,9 +4,20 @@ class Monoid m where
...
+
+class ForMonoid private constructor() { companion object }
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]: Even though, it's directly from the arrow doc's the reader won't be familiar with either of the following type classes or the reason, why we use MonoidOf. 
Just delete the lines 9 - 11 and 
remove Semigroup<A>, MonoidOf<A> and the combineAll functions.

File: src/main/ank/2.3-free-monoids.md [hunk 7]
```diff
@@ -4,9 +4,20 @@ class Monoid m where
...
+typealias MonoidOf<A> = arrow.Kind<ForMonoid, A>
+fun <A> MonoidOf<A>.fix(): Monoid<A> = this as Monoid<A>
+
+interface Monoid<A> : Semigroup<A>, MonoidOf<A> {
+  fun empty(): A
+
+  fun Collection<A>.combineAll(): A =
+    if (isEmpty()) empty() else reduce { a, b -> a.combine(b) }
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]: Instead, write the combine function explicitly.

File: src/main/ank/2.3-free-monoids.md [hunk 8]
```diff
@@ -15,10 +26,35 @@ instance Monoid [a] where
   mempty = []
   mappend = (++)
 ```
-```kotlin:ank:silent
-class ListMonoid<M> : Monoid<List<M>> {
-  override fun mempty() = listOf<M>()
-  override fun mappend(a1: List<M>, a2: List<M>) = a1 + a2
+```kotlin
+package free
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]: remove package declaration

File: src/main/ank/2.3-free-monoids.md [hunk 9]
```diff
@@ -28,7 +64,7 @@ class ListMonoid<M> : Monoid<List<M>> {
 ```
 ```kotlin
 2 * 3 = 6
-listOf(2) + listOf(3) == listOf(2, 3) // not the same as listOf(6)
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]: keep it simple. revert the changes to 
> listOf(2) + listOf(3) == listOf(2, 3) // not the same as listOf(6)

File: src/main/ank/2.3-free-monoids.md [hunk 10]
```diff
@@ -44,7 +80,7 @@ h(a * b) == h(a) * h(b)
 [2] ++ [3] = [2, 3]
 ```
 ```kotlin
-listOf(2) + listOf(3) == listOf(2, 3)
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]: here the same as above

File: src/main/ank/2.3-free-monoids.md (hunk 8 above)
Comment (nomisRev) [reply src/main/ank/2.3-free-monoids.md]:
```suggestion
```

File: src/main/ank/2.3-free-monoids.md (hunk 9 above)
Comment (nomisRev) [reply src/main/ank/2.3-free-monoids.md]:
```suggestion
listOf(2) + listOf(3) == listOf(2, 3) // not the same as listOf(6)
```

File: src/main/ank/2.3-free-monoids.md (hunk 10 above)
Comment (nomisRev) [reply src/main/ank/2.3-free-monoids.md]:
```suggestion
listOf(2) + listOf(3) == listOf(2, 3)
```

File: src/main/ank/2.3-free-monoids.md (hunk 5 above)
Comment (nomisRev) [reply src/main/ank/2.3-free-monoids.md]:
```suggestion
```

Comment (nomisRev) [reply src/main/ank/2.3-free-monoids.md]:
```suggestion
```

File: src/main/ank/2.3-free-monoids.md (hunk 8 above)
Comment (nomisRev) [reply src/main/ank/2.3-free-monoids.md]:
```suggestion
package free
```

File: src/main/ank/2.3-free-monoids.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+  mempty = []
+  mappend = (++)
+```
+```kotlin
+package free
+
+import arrow.higherkind
+import arrow.typeclasses.Monoid
+
+operator fun <A>FreeMonoid<A>.plus(b: FreeMonoid<A>) = FreeMonoid.just(this.terms + b.terms)
```
Comment (nomisRev) [inline src/main/ank/2.3-free-monoids.md]:
```suggestion
operator fun <A> FreeMonoid<A>.plus(b: FreeMonoid<A>) = FreeMonoid.just(this.terms + b.terms)
```

File: src/main/ank/2.3-free-monoids.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+@higherkind
+sealed class FreeMonoid<S>(val terms: List<S>): FreeMonoidOf<S> {
+
+    internal data class Pure<A>(val a: List<A>) : FreeMonoid<A>(a)
```
Comment (nomisRev) [inline src/main/ank/2.3-free-monoids.md]: Any reason for this to be an ADT? Can't we do `class FreeMonoid<S>(val terms: List<S>): FreeMonoidOf<S>` instead?

File: src/main/ank/2.3-free-monoids.md (hunk 2 above)
Comment (nomisRev) [reply src/main/ank/2.3-free-monoids.md]:
```suggestion
```kotlin
```

File: src/main/ank/2.3-free-monoids.md (hunk 12 above)
Comment (i-walker) [reply src/main/ank/2.3-free-monoids.md]: I suggest you @arioston check both Links to integrate the free monoid better with Arrow. 
- https://medium.com/@olxc/free-monads-explained-pt-1-a5c45fbdac30
- https://typelevel.org/cats/datatypes/freemonad.html

Feel free to ask questions.

File: src/main/ank/2.3-free-monoids.md [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+}
+```
+................
+```Haskell
+instance Monoid [a] where
+  mempty = []
+  mappend = (++)
+```
+```kotlin
+import arrow.higherkind
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]: This is ideally, what you're trying to do
`
@extension
interface Monoid<A>: MonoidOf<List<A>>{
  fun empty() = emptyList()
  fun List<A>.combine(other: List<A>) = this + other
}
`

Comment (i-walker) [reply src/main/ank/2.3-free-monoids.md]: These are the imports
```
import arrow.extension
import arrow.typeclasses.MonoidOf
```

File: src/main/ank/2.3-free-monoids.md [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+  fun combine(a: A, b: A): A
+}
+```
+................
+```Haskell
+instance Monoid [a] where
+  mempty = []
+  mappend = (++)
+```
+```kotlin
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]:
```suggestion
```kotlin:ank:playground
```

File: src/main/ank/2.3-free-monoids.md [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+import arrow.extension
+import arrow.typeclasses.MonoidOf
+
+@extension interface Monoid<A>: MonoidOf<List<A>> {
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]:
```suggestion
@extension 
interface Monoid<A>: MonoidOf<List<A>> {
```

File: src/main/ank/2.3-free-monoids.md [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+```
+```kotlin
+2 * 3 == 6
+```
+................
+```Haskell
+p :: x -> U m
+```
+```kotlin
+val p : (X) -> Kind<U, M>
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]:
```
from the book:
We start with a set of generators, x. That's a set in Set. 
The pattern we are going to match consists of monoid m - an object of Mon - and a function p in Set,
where U is our forgetful functor from Mon to Set. 
This is a weird heterogeneous pattern - half in Mon and half in Set.
The idea is that the function p will identify the set of generators inside the X-ray image of m. 
(Page 212)
```
@nomisRev: Adjusted to his response
```suggestion
fun <A, U, M> p(x: A): Kind<U, M>
```

File: src/main/ank/2.3-free-monoids.md [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+instance Monoid [a] where
+  mempty = []
+  mappend = (++)
+```
+```kotlin:ank:playground
+import arrow.extension
+import arrow.typeclasses.MonoidOf
+
+@extension 
+interface Monoid<A>: MonoidOf<List<A>> {
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md:23]:
```suggestion
interface Monoid<A>: MonoidOf<ForListK> {
```

File: src/main/ank/2.3-free-monoids.md [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+}
+```
+................
+```Haskell
+instance Monoid [a] where
+  mempty = []
+  mappend = (++)
+```
+```kotlin:ank:playground
+import arrow.extension
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md:19]:
```suggestion
import arrow.data.ForListK
import arrow.extension
```

File: src/main/ank/2.3-free-monoids.md (hunk 17 above)
Comment (pakoito) [reply src/main/ank/2.3-free-monoids.md:23]: That shouldn't be the case.

Comment (i-walker) [reply src/main/ank/2.3-free-monoids.md:23]: I thought ...Of<HKT> expects a HKT?
At least, that is what I've been said.

Comment (pakoito) [reply src/main/ank/2.3-free-monoids.md:23]: Monoid is aggregation of regular values, check the existing implementation of `ListK`.

File: src/main/ank/2.3-free-monoids.md (hunk 16 above)
Comment (nomisRev) [reply src/main/ank/2.3-free-monoids.md]: This is actually not valid Kotlin code.

```
fun <X, U, M> p(a: A): Kind<U, M>
```

or

```
interface Example<X, U, M> {
   val p : (X) -> Kind<U, M>
}
```

There is no way to use generics unless they are defined in a function, or the lambda lives inside a class or interface that defines those generics.

File: src/main/ank/2.3-free-monoids.md [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+```kotlin
+val p : (X) -> Kind<U, M>
+```
+
+................
+```Haskell
+q :: x -> U n
+```
+```kotlin
+val q: (X) -> Kind<U, N>
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]:
```suggestion
fun <A, U, N> q(x: A): Kind<U, N>
```

File: src/main/ank/2.3-free-monoids.md [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+```kotlin
+val h: (M) -> N
+```
+
+................
+```Haskell
+q = U(h) . p
+```
+```kotlin
+val q = uh compose p
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]:
```suggestion
val q = `Uh` compose p
```

File: src/main/ank/2.3-free-monoids.md [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+```kotlin
+val q: (X) -> Kind<U, N>
+```
+
+................
+```Haskell
+h :: m -> n
+```
+```kotlin
+val h: (M) -> N
```
Comment (i-walker) [inline src/main/ank/2.3-free-monoids.md]:
```suggestion
fun <M, N> h(m: M): N
```

**Comment by i-walker** [comment]
😊😊🥳 Thank you good Job. 💪🏾
BTW you're one of my first "reviewees". I hope you liked it.
